markdownlint/TOP003_defaultSectionContent/TOP003_defaultSectionContent.js

```diff
@@ -27,92 +28,144 @@ function getListSectionErrors(sectionTokens, section) {
   );
 
-  const hasDefaultContent =
-    tokensAfterHeading[0].line === listSectionsDefaultContent[section];
-  if (!hasDefaultContent && tokensAfterHeading[0].type !== "bullet_list_open") {
+  const listItemTokens = tokensAfterHeading.filter(
+    (token) => token.type === "list_item_open"
+  );
+  const nestedListItemTokens = listItemTokens.filter(
+    (token) => token.level > 1
+  );
+  nestedListItemTokens.forEach((nestedListItemToken) => {
     listSectionErrors.push(
+      // We're not applying a fix because we can't always know whether just un-nesting the list
+      // will resolve the issue. There may be entire list items that need to be removed as well.
       createErrorObject(
-        tokensAfterHeading[0].lineNumber,
-        `Expected: "${listSectionsDefaultContent[section]}"; Actual: "${tokensAfterHeading[0].line}"`
+        nestedListItemToken.lineNumber,
+        `The ${section} section must not contain nested lists.`
       )
     );
-  }
-  if (!hasDefaultContent && tokensAfterHeading[0].type === "bullet_list_open") {
+  });
+
+  // https://regexr.com/80hjf to test following regex. Note that we don't test nested lists due to
+  // the fact that whitespaces won't precede the token marker, e.g. "1.", in the token object
+  const orderedListItemRegex = /^\d+\.\s*/;
+  const orderedListItemTokens = listItemTokens.filter((token) =>
+    orderedListItemRegex.test(token.line)
+  );
+  orderedListItemTokens.forEach((orderedListItemToken) => {
     listSectionErrors.push(
       createErrorObject(
-        tokensAfterHeading[0].lineNumber,
-        `Expected default section content to precede the unordered list of ${listItemsName}: "${listSectionsDefaultContent[section]}"`
+        orderedListItemToken.lineNumber,
+        `The ${section} section must not include any ordered lists.`,
+        {
+          lineNumber: orderedListItemToken.lineNumber,
+          deleteCount:
+            orderedListItemToken.line.match(orderedListItemRegex)[0].length,
+          insertText: "- ",
+        }
+      )
+    );
+  });
+
+  const defaultContentOpenTokenIndex = tokensAfterHeading.findIndex(
+    (token) => token.line === listSectionsDefaultContent[section]
+  );
+
+  if (defaultContentOpenTokenIndex > 0) {
+    const defaultContentToken =
+      tokensAfterHeading[defaultContentOpenTokenIndex];
+    listSectionErrors.push(
+      createErrorObject(
+        defaultContentToken.lineNumber,
+        `Expected default section content to come immediately after the ${section} heading.`
       )
     );
   }
+  if (defaultContentOpenTokenIndex === -1) {
+    const errorDetail = `Expected: "${listSectionsDefaultContent[section]}"; Actual: "${tokensAfterHeading[0].line}",`;
+
+    listSectionErrors.push(
+      createErrorObject(tokensAfterHeading[0].lineNumber, errorDetail, {
+        lineNumber: tokensAfterHeading[0].lineNumber,
+        insertText: `${listSectionsDefaultContent[section]}\n\n`,
```

Comment:
Currently, the correct default content is inserted but the incorrect content is left there, meaning the user would still have to manually delete it.

With the below, I've tried to think of edge cases that could break *and* cause huge problems because of this but I can't think of any. As far as I'm aware, the error with this fixInfo would only throw when the first item isn't the default content *and* the correct default content isn't present in the section.
(Also why the template string for insertText? :P)

```js
  lineNumber: tokensAfterHeading[0].lineNumber,
  deleteCount: tokensAfterHeading[0].line.length,
  insertText: listSectionsDefaultContent[section],
```

The only situation where this could cause issues would be if the very first item *was* the unordered list, in which case the above would just replace that first UL item.

A hacky solution could be to check if the offending first item starts with `- ` then insert `` `${listSectionsDefaultContent[section]}\n\n${tokensAfterHeading[0].line}`  ``
e.g.

```js
  if (defaultContentOpenTokenIndex === -1) {
    const errorDetail = `Expected: "${listSectionsDefaultContent[section]}"; Actual: "${tokensAfterHeading[0].line}",`;
    let insertedText = listSectionsDefaultContent[section];

    if (tokensAfterHeading[0].line.startsWith("- ")) {
        insertedText += `\n\n${tokensAfterHeading[0].line}`;
    }

    listSectionErrors.push(
      createErrorObject(tokensAfterHeading[0].lineNumber, errorDetail, {
        lineNumber: tokensAfterHeading[0].lineNumber,
        deleteCount: tokensAfterHeading[0].line.length,
        insertText: `${listSectionsDefaultContent[section]}`,
      })
    );
  }
```

The above will solve that edge case and doesn't seem to break any others from my testing.

Reply:
Yeah I wasn't sure whether we should remove content and then insert content for this one, just because we may not know what exactly is there to possibly remove. I opted to just insert the default content, then any subsequent runs of the fixer should catch/fix the "additional content between default and list" portion.

If you feel that there shouldn't be any issues with something like the above I can add that change, though, let me know what you think.

Reply:
I think it should be fine.

If it goes
```md
### Heading

Unwanted stuff (inc. ul)

Correct default text
```
then the `Expected default section content to come immediately after the ${section} heading.` error gets flagged instead.

If the correct default text is not there, then the only instance of the first line being something we want would be if the UL is the first item, in which case the 2nd snippet I provided will cover that (for LO/KC/AR sections).

And any other edge case I can think of would result in bigger problems anyway and surely would be caught in review if not by the author (and likely would leave post-fix errors anyway).

Doing the delete + insert would make most realistic situations handled in 1 pass, perhaps 2 just for any extra blank lines left behind that we can't address.